File: backtracking/prob22.py
# 22 generate parenthesis

class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """

        ans = []

        def bt(s, left, right):
            if len(s) == 2 * n:
                ans.append(s)
                return

            if left < n:
                bt(s+"(", left+1, right)
            if left > right:
                bt(s+")", left, right+1)

        bt("", 0, 0)

        return ans

File: interview/prob22.go
func generateParenthesis(n int) []string {
    
    ans := []string{}
    
    walk("", n, 0, 0, &ans)
    
    return ans
    
}

func walk(s string, n int, left int, right int, ans *[]string) {
    if left == n && right == n {
        *ans = append(*ans, s)
    }
    
    if left < n {
        walk(s + "(", n, left+1, right, ans)
    }
    
    if right < left {
        walk(s + ")", n, left, right+1, ans)
    }
    
}
